### ISSUE TITLE
Keyset Persistence

### ISSUE THREAD

**Issue by dbnoble** [froatsnook/meteor-shopify#9]
Why are we persisting the keysets in memory instead of simply storing them in a collection on the database? 

Especially since there is no way to check if a current keyset exists, the keyset has to be reestablished constantly to be fully reliable. Every time you push code, the keysets will all be flushed.

**Comment by froatsnook** [comment]
The intention was to leave this up to the library user.

I have two apps in production, and each is handled differently:

One uses private API, and there I store the credentials in `settings.json` and call `Shopify.addKeyset` in `Meteor.startup` on the server.

The other uses the public API.  This one has multiple users per client (with one shop per client).  So after a user authenticates, the credentials are stored in the Clients collection, and then I memoize the shopify API object for each client:

``` javascript
// server/lib/ShopifyGetAPI.js

// Cache API objects per shop.
//
// shop => Shopify.API
var apis = { };

// Get Shopify API access object for a client.
//
// @param clientId {String} _id of client.
ShopifyGetAPI = function(clientId) {
    check(clientId, NonEmptyString);
    var client = Clients.findOne(clientId);

    var shop = client.services.shopify.shop;
    var access_token = client.services.shopify.access_token;
    check(shop, String);
    check(access_token, String);

    if (!apis[shop]) {
        apis[shop] = new Shopify.API({
            shop: shop,
            access_token: access_token
        });
    }

    return apis[shop];
};
```

In other words, the first time you start using `Shopify.API` for a client, it gets the credentials from the database.  Since the `access_token` is provided directly, creating a keyset manually isn't necessary.

Does that make sense?  Would something like this be sufficient for you?  It would be easy to add something like `Shopify.keysetExists` (as a workaround for now you could do something like `keyset in Shopify._keysets`).  I'm not sure how I feel about automatically putting them in the database.  But `Shopify.persistKeysets = true` sounds doable!

**Comment by dbnoble** [comment]
I guess my main problem stems from me not fully understanding the intended authorization flow for getting the auth token saved and saving it only on the SERVER because the example app (and all of the examples) seem to be doing this client side where in production they should be handled server-side. 

I think for the most part this is just the frustration of so many hours working on this that it's just not "clicking" in my head. I feel like most of this could be solved fairly easily by simply specifying my own redirect_uri, but I can't wrap my head around how to make that work without getting the "no handler" errors.

Currently I am using a Router.beforeAction to check for an existing user with the shop parameter, or if no user is found, passing the auth token through onAuth to a method call that creates a user via Accounts.updateOrCreateUserFromExternalService and generates the keyset. 

Is the proper, secure way to do it?

I feel like the access token should not be being passed through onAuth since it is clientside, am I wrong on that?

How do you suggest identifying which user to get credentials for? Currently I am generating keysets using the 'shop' parameter as the name and grabbing the auth token from the user's service configuration on the server where Meteor.user.findOne({shop: shop}). Then on the client I can grab the shop param and specify that variable as the keyset to use. This is obviously ridiculous because one could just pass an arbitrary shop parameter and get access through their keyset.

**Comment by dbnoble** [comment]
Also I should note that I was initially actually logging the user in, and in the onBeforeAction doing a check to see if the current user's shop matched the loaded shop param and send them back through reauth and login if not (my target demographic is likely to have multiple shops). I didn't really like the way that worked so I decided to just use the accounts for the OAuth service config and not actually log users in. The downside is that I can't rely on checking against the current userId to validate the shop parameter.

**Comment by froatsnook** [comment]
I feel your pain. Originally this library was just private API and much simpler. 

The public app workflow (with PublicAppOAuthAuthenticator.openAuthTab)  is a combination of client (by necessity) and sever (for security). Your secrets live on the server and the client just starts the Auth process. The Auth finishes on the server, so the client never even needs to see his access_token. I can live in a named keyset, which is how you can use Shopify.API objects on the client. I hope you can get it working since it's definitely the easy and tested solution. 

Specifying your own redirect_uri is not something I've ever done, so if you're getting "no handler" errors, possibly there's just a small bug somewhere in the framework. 

I am not sure how meteor login through external service works. My users have passwords, which is how I identify them. Let me get back to you on that.

**Comment by froatsnook** [comment]
I am currently out of the office (and I've never used these features so I'm not positive it works like this), but I would aim for this:
1. Users create an account by logging in via some shopify shop. Save the service name as the shop name. This saves the access token in user.services and sets a cooki so that the user remains logged in. 
2. The user can add additional shops somewhere in an account page by opening another Auth tab. Again save shop name as the service name so that user.services has credentials for both shops. 
3. Now the user should be able to login by specifying either shop name and opening an Auth tab, right? 
4. Make a meteor method called GetKeyset which takes a shop parameter. This should get the access_token from user.services, then return the name of the keyset for that shop (creating if necessary). Use a UUID for the keyset name, and have a lookup table on the server of shop name -> keyset name. 

If external service login doesn't work like that, then do all the above except with password login (this is how my app works). Just don't put the access_tokens in profile and then it will be server side.

**Comment by dbnoble** [comment]
Thanks for your response and work on the package.

I did consider that I could still log the user in and just add the different shops as an array instead of a string, so that's something I may work on, although it might get rather complex in terms of creating charges through Shopify's billing API lol.

Primarily I just wanted to know if passing the access token through onAuth and then calling a method was actually the proper secure way to do it. The reason I ask, I tested using console.log inside onAuth and found that it executes on the CLIENT and not the server, so I am assuming that anything I put inside onAuth is only being executed client side. 

Am I supposed to be defining authenticator on both the server and client? I have only had it as client-side thus far.

**Comment by froatsnook** [comment]
Yes you're right. Based on how I'm thinking about it now, it's insecure since the access_token is leaking to the client. When I wrote this code I just didn't want to expose the secret keys. But I think it would be better to not expose it. I will add secure:true as an option and then figure out a way to do onAuth server side. Possibly you can pass a value userData which will given to onAuth callbacks installed server side:

```
Shopify.onAuth(function(access_token, userData) {...});
```

**Comment by dbnoble** [comment]
Okay I've made some headway on all fronts using my own redirect_uri and handling the code conversion and post-auth functions server-side.

Here is a basic example:

client-side authenticator:

```
// Just a generic example, change to suit your needs

// Grab the shop parameter from the request params if needed
var params = Router.current().params.query;
var $shopName = params.shop.replace('.myshopify.com', '');  

//define authenticator
var authenticator = new Shopify.PublicAppOAuthAuthenticator({
  shop: $shopName,
  api_key: 'YOUR_API_KEY',
  keyset: "default",
  scopes: "read_products,write_products,read_script_tags,write_script_tags,read_orders,write_orders",
  // IMPORTANT: Specify your custom redirect_uri instead of using the default
  redirect_uri: Meteor.absoluteUrl("__shopify-custom-auth")
  // We don't need onAuth anymore since we'll handle it all on the server
});

// When you are ready to begin the auth flow, call your desired method
// Don't forget to use window.top.location.href instead if calling from inside embedded app
location.href = authenticator.auth_uri;
```

/server/init.js

```
// Create an object resembling your client-side PublicAppOAuthAuthenticator.options
// since some PublicAppOAuthAuthenticator methods are not available on the server
myAuthenticatorConfig = {
  keyset: 'default',
  api_key: 'YOUR_API_KEY',
  // Specify a URL to return to after auth token is requested, in this case we'll append the shop parameter
  // after auth so we can send the user directly back to our embedded app (see notes in routes.js) 
  after_auth_uri: '.myshopify.com/admin/apps/YOUR_API_KEY'
}
```

/lib/routing/routes.js

```
// Define your custom auth redirect_uri as a server side route using iron:router
Router.map(function () {
  this.route('__shopify-custom-auth', {
    path: '/__shopify-custom-auth',
    // This designates the route as a server side route
    where: 'server',
    action: function () {
      // Grab the shop and code params from the request
      var params = this.params.query;
      var shop = params.shop.replace('.myshopify.com', '');
      var code = params.code;
      // We couldn't have known the shop when we defined the myAuthenticatorConfig
      // object in /server/init.js, so we'll add it to the object here
      myAuthenticatorConfig.shop = shop;
      // Call Shopify._getPermanentAccessToken directly since the clent side authenticator instance and
     // authenticator.getPermanentAccessToken's methods are not available on the server
      Shopify._getPermanentAccessToken(myAuthenticatorConfig, code, function(err, access_token){
        if(!err){
          // Move any operations that were previously in the onAuth callback here
          // You can do anything you need with the access token here, create user, generate keyset, etc
        } else {
          console.log(err);
        }
      });
      // Redirect back to your app (assuming you are using the Embedded App SDK)
      this.response.writeHead(302, {
        'Location': 'https://' + shop + myAuthenticatorConfig.after_auth_uri
      });
      this.response.end();
    }
  }); 
});
```

Works perfectly, although I recommend implementing HMAC verification on a production app, which will also be useful when identifying which keyset to use.

**Comment by froatsnook** [comment]
Hey @dbnoble.  I'm glad you were able to get it working.  Do you have any suggestions for docs or code to make this easier?

I had a look at the code and saw that I started work on a 1.5.0 branch specifically to deal with the access_token being leaked to the client.  I'll publish this once I've written some tests.

Would having a server-side onAuth handler that looks like this fix your issues with the default redirect_uri?  Then the client-side onAuth wouldn't receive the access_token.

``` javascript
// server/init.js
Shopify.onAuth(function(access_token, authConfig, userId) {
    ...
});
```

I'm thinking HMAC verification will also be implemented for 1.5.0.

**Comment by dbnoble** [comment]
@froatsnook a serverSide onAuth hook would be excellent! Paired with persisted keysets, that would allow us to use a waitOn for the app's route to prevent the race condition I have experienced where the template renders before the keyset exists, causing the API to init with undefined credentials.

Is there any way to to use the default handler without opening a new tab/window? I've noticed that at least Chrome blocks it by default and you have to manually allow popups. Just trying to prevent silly support requests.

**Comment by froatsnook** [comment]
Hey @dbnoble.  So are you thinking you'll subscribe to the Keysets in the waitOn?  How does a server-side onAuth help you in this scenario?  Maybe you could post some "sample code" of how it should work? 

As for the new tab, somehow the user has to get to shopify's login page.  There are 3 options I thought of and listed in the readme.  The first is opening a new tab, which you know.  The second is [opening an iframe](https://github.com/froatsnook/meteor-shopify#scenario-2-iframe), and the 3rd is [directly navigating](https://github.com/froatsnook/meteor-shopify#scenario-3-direct-navigation).  Does one of those sound good?  Let me know if something doesn't work.  These scenarios haven't seen much testing.

It might even be better if you create separate issues!  That way if someone else has the same question, they might find the answer more easily.

**Comment by froatsnook** [comment]
How are your keysets created?

**Comment by froatsnook** [comment]
1.5.0 is published with server-side `onAuth`, `Shopify.harden()` to not leak `access_token`.  No HMAC or keyset persistence yet.

**Comment by dbnoble** [comment]
@froatsnook I have moved back to logging in users for the time being by using a server-side 'gateway' page and if the current shop param does not match the current logged in user's shop it logs them out and reauthenticates. It will then log in the appropriate user or creates a new user if necessary after validating the HMAC. This works and is more or less instant when switching users, and aside from HMAC verification, it is bypassed completely as long as the current user's shop matches the requesting shop.

That makes it more secure by letting Shopify's oauth flow confirm that the user is valid and has access to the resources that belong to that shop's user, instead of using the shop parameter alone as an identifier.

I am also now naming the keyset to match the userId since there is no way to access userIds externally, and the logged in user only has access to their own ID. This prevents anyone from hijacking the keyset by simply passing in a different shop as the keyset name.

Keyset generation is done during user creation server-side using your existing function as follows:

```
Shopify.addKeyset(userId, {
             access_token: access_token
            });
```

I also upsert the details to a Keysets collection in the DB:

```
Keysets.upsert({name: result.userId}, {$set: { token: access_token }});
```

At startup, the server loops through the collection and reloads them so that they will persist between server restarts or when the cluster is scaled up and new servers added.

```
Meteor.startup(function () {
  existingKeysets = Keysets.find({});
  existingKeysets.forEach(function(keyset) {
     Shopify.addKeyset(keyset.name, {
     access_token: keyset.token
     });
  });
```

HMAC verification via the following code based on @lukeclifton 's meteor-shopify-api package:

```
Meteor.methods({
    'validateSignature': function(params) {

        check(params, Object);

        var hmac = params.hmac;

        // Delete signature and hmac as shopify docs specifies
        delete params.signature;
        delete params.hmac;

        // Create message query string
        var message = serializeObject(params);

        // Do the hmac sha256 encrypting
        var hash = CryptoJS.HmacSHA256(message, ShopifyApi.options.secret).toString();

        // Return true if we have a match, otherwise return false
        return hash === hmac;
    }
});
```

I have not migrated to 1.5.0 branch yet.

**Comment by dbnoble** [comment]
In regards to using waitOn for routes, now that I have moved everything back to a logged-in user environment, I can simply waitOn the user details publication on my route so that the dashboard will not render and try to instantiate the API before the userId is available after creation.

I should note that I have never had to do that before. For some reason, the user object does not seem to be available as early as in other projects I have worked on, so the page was being rendered and API init was happening before the user object was available. I was also experiencing the same problem with keysets, where the page was being rendered prior to the keyset method call completing, so the api request would fail the first time since the keyset did not actually exist at init.

Server:

```
Meteor.publish('myProfile', function() {
  return Meteor.users.find(this.userId, {fields: {profile: 1 }});
});
```

In applicable routes:

```
waitOn: function(){
      return Meteor.subscribe('myProfile');
    }
```

My initial idea of using waitOn was based on using DB persisted keysets and I could simply publish the keyset corresponding with the current shop then use waitOn to make sure it was loaded prior to the page rendering. Now that everything utilizes data stored in the user object, I can just waitOn that instead.

**Comment by froatsnook** [comment]
OK, thanks for the info!

Did I understand this correctly?  You're using Shopify's auth for account creation and login?  That sounds really cool.  Are you still using PublicAppOAuthAuthenticator?  Did you use iframes or direct login or opening a new tab or something else?  How does the "gateway" page work?

I also have never had problems accessing the user object.  But I think it's because I have something like this in my base controller:

``` javascript
    onBeforeAction: function() {
        if (Meteor.loggingIn()) {
            this.render("loading");
        } else if (!Meteor.user()) {
            this.layout("empty");
            this.render("login");
        } else {
            this.next();
        }
    },
```

Since you're doing the keyset creating server-side, the migration to 1.5.0 should be as easy as putting `Shopify.harden()` in server-side `Meteor.startup`.

**Comment by lukeclifton** [comment]
Need to catch up on this chat but in the meantime, are you guys both still using Iron Router?

**Comment by froatsnook** [comment]
@lukeclifton Yes, I'm using iron router.

**Comment by dbnoble** [comment]
@lukeclifton Yep, still using iron router.

@froatsnook in terms of the gateway, I guess page is the wrong word, it's actually a clientside route that does some checks and either sends the user to the dashboard or redirects them to the authentication flow similar to the one I posted earlier.

I am indeed still using PublicAppOauthAuthenticator with direct navigation and a slightly different version of the server-side redirect_uri handler that I posted previously.

Using the accounts-base and service-configuration, you can use any oauth provider for user creation/login just like "Login with facebook", "Login with twitter", etc, although you have to write your own handler for services that don't already have one. Since @lukeclifton had already done some work in that area I based my strategy on code from his package to handle some of the operations.

I should note that this would probably be handled better in an onBeforeAction, but I think that the onBeforeAction that you referenced earlier was short circuiting mine and causing it to go into a redirect loop during login, so I just made it a hard route and set that as the primary app url in Shopify's partner dashboard.

Here is the code for the flow I am currently using. Please note that I'm still experimenting with everything to get the desired result, so I have alot of console.log's in there as sanity checks and I'm self-taught and do things a bit goofy sometimes. I'm expecting to go back in and clean things up a bit once the ACTUAL APP functionality is finished.

```
Router.route('entryGate', {
    path: '/entryGate',
    where: 'client',
    loadingTemplate: 'loading',
    //
    // Make sure user details are available before executing action
    //
    waitOn: function(){
      return Meteor.subscribe('myProfile');
    },
    action: function () {
    var params = this.params.query;
    var shop = params.shop.replace('.myshopify.com', '');
    console.log('Loaded shop:' + shop);
    var failedCheck = false;
    if(Meteor.user()){
      console.log('Current User shop:' + Meteor.user().profile.shop);
     if(Meteor.user().profile.shop !== shop){
       console.log('Shops do not match!');
      failedCheck = true;
       Meteor.logout();
      } else {
        console.log('Shops appear to match!');
      }
    }
    if (!Meteor.user() || failedCheck) {
            // Prepare to authenticate
            var authenticator = new Shopify.PublicAppOAuthAuthenticator({
                shop: shop,
                api_key: 'YOUR_API_KEY',
                keyset: 'default',
                scopes: 'read_products,write_products,read_script_tags,write_script_tags', 
                embedded_app_sdk: true,
                redirect_uri: Meteor.absoluteUrl('__shopify-custom-auth')
            });

            // Start authenticating - MUST use window.top here if using embedded SDK
            window.top.location.href = authenticator.auth_uri;

    } else {
      // User's recorded shop matches, just send them straight to the dashboard
      Router.go('/dashboard?shop=' + shop);
    }
  }
});
```

This is the updateOrCreateUser method we will be using in the handler, thanks @lukeclifton for this

```
'updateOrCreateUser': function(shop, accessToken) {

        check(shop, String);
        check(accessToken, String);

        // Get shop name from shop
        var shopName = shop.replace('.myshopify.com', '');

        var serviceData = {
            id: shop,
            accessToken: accessToken,
            shopName: shopName,
            shop: shop
        };

        // Accounts.updateOrCreateUserFromExternalService is a function located in accounts-base/accounts_server.js
        // This function updates or creates a user with the external service (shopify) authentication data specified above.
        return Accounts.updateOrCreateUserFromExternalService('shopify', serviceData);
    },
```

I also am using the onCreateUser hook to write the user's shop to the profile since the user's profile object is already made available to a logged in user.

```
Accounts.onCreateUser(function(options, user) {
    user.profile.shop = user.services.shopify.shop;
  return user;
});
```

I have not pushed my revision with HMAC verification to my repo so I don't have it available at the moment, but I already posted that above. Server-side route:

```
Router.route('__shopify-custom-auth', {
    path: '/__shopify-custom-auth',
    where: 'server',
    action: function () {
      var params = this.params.query;
      var shop = params.shop.replace('.myshopify.com', '');
      var code = params.code;
      var thisRequest = this;
      var endRes = function(thisRequest, result){
        thisRequest.response.writeHead(302, {
        'Location': '/login/' + result.userId + '/' + shop
        });
        thisRequest.response.end();
      };
      var failedRes = function(thisRequest){
      var json = '{"status": "failed"}';
      thisRequest.response.setHeader('Access-Control-Allow-Origin', '*');
      thisRequest.response.setHeader('Content-Type', 'application/json');
      thisRequest.response.end(json);
      };
      myAuthenticatorConfig.shop = shop;
      Shopify._getPermanentAccessToken(myAuthenticatorConfig, code, function(err, access_token){
        if(!err){
           Meteor.call('updateOrCreateUser', shop, access_token, function(error, result){
             Shopify.addKeyset(result.userId, {
             access_token: access_token
            });
             Keysets.upsert({name: result.userId}, {$set: { token: access_token }});
             endRes(thisRequest, result);
           });
        } else {
          console.log(err);
          failedRes(thisRequest);
        }
      });
    }
  });
```

Since the redirect_uri handler is a server-side route, we send the user to a client-side route at /login/USERNAME/SHOP so we can actually use the loginWithShopify function. I am also passing the shop to it so that I can send the user directly back to the app embedded in the Shopify admin afterwards.

```
Router.route('/login/:userId/:shop', function(){
  loginWithShopify(this.params.userId, this.params.shop);
});
```

The loginWithShopify function is defined in the client-side /client/init.js
Based off @lukeclifton's code again, modified for the redirect back to embedded app after login

```
loginWithShopify = function(userId, shop) {

    check(userId, String);

    console.log('Shopify app: OAuth process completed. Logging in...');

    // create a login request with shopify: true, so our custom loginHandler can handle this request
    var loginRequest = {
        shopify: true, 
        userId: userId
    };

    Accounts.callLoginMethod({
        methodArguments: [loginRequest],

        // login success callback
        userCallback: function(error) {
            if (!error) {

                console.log('Shopify app: Logged in successfully');

            // Use a server route redirect here seems to be faster than using window.location
            Router.go('/loadApp/' + shop);
            } else {
                console.log(error);
            }
        }
    });
};
```

And finally here is the loadApp route (server-side redirect)

```
Router.map(function () {
  this.route('loadApp', {
    path: '/loadApp/:shop',
    where: 'server',
    action: function () {
      this.response.writeHead(302, {
        //
        // the after_auth_uri would be '.myshopify.com/admin/apps/YOURAPIKEY'
        // Please note that this is the SERVER-SIDE facsimile of the authenticator
        'Location': 'https://' + this.params.shop + myAuthenticatorConfig.after_auth_uri
        });
        this.response.end();
    }
  }); 
```

I hope that all makes sense. Any questions feel free to ask!

**Comment by lukeclifton** [comment]
@dbnoble Glad you found some of the code from my original meteor-shopify-api package useful. I needed to get logging in via Shopify OAuth / custom external service working from the get go so took me a while but got there in the end. 

Its probably not perfect but points in the right direction. It was my first Meteor package.

Its my intention to re-write the package now that i have a lot more experience with Meteor / Shopify API. With the the following objectives:
- Try and structure the package to match the official OAuth packages a little closer
- Stop depending on IronRouter (as its now no longer maintained) and move to FlowRouter
- Ensure its compatible and makes use of all new features of Meteor 1.2

Maybe rather than rewriting my package, i could just contribute to this package but i need to take a closer look and test drive it to ensure it gives me everything i need for my current projects.

**Comment by froatsnook** [comment]
@dbnoble Fantastic.  Thanks a lot for posting the code.  It'd be nice if the package could do all the heavy lifting for you.  I'll have to think about this some more.

@lukeclifton Contributions would be much appreciated!  To give you an idea of where this package stands w.r.t. your objectives, I'm pleased that the OAuth is working for people, but it definitely doesn't match the official OAuth packages.  Integrating it with accounts somehow like dbnoble did in his app sounds like an interesting approach. There's no dependency on Iron Router (maybe my previous comment misled you). As for Meteor 1.2, I'd at least like to start using the ecmascript package.  I actually haven't upgraded my app in production to 1.2 yet, so I'm not sure if there will be compatibility issues.  I hope not...

**Comment by dbnoble** [comment]
Yes obviously my implementation is quite hack-y but this is almost by necessity. My target market for the app I am working on is significantly more likely to be running multiple Shopify stores, so I really have to take that into account and provide some type of logic for switching users since Meteor logins are persisted in local storage. 

Ideally I could just share the shop and access tokens across stores on a single user but I simply don't have time to build out that type of functionality within my timeline.

Now that I have something that more or less works for what I need, I'm going to shift my focus on building the actual app and worry about making the login system perfect later.

**Comment by lukeclifton** [comment]
@dbnoble yeah i had to take the same kind of approach, get it working with the plan of coming back later to perfect. 

Out of curiosity where abouts are you guys based? @dbnoble @froatsnook

**Comment by froatsnook** [comment]
@dbnoble: Good luck with the app!

**Comment by froatsnook** [comment]
@dbnoble, @lukeclifton: Is it true that HMAC verification is only used for webhooks?  So far this package doesn't deal with webhooks at all.

**Comment by lukeclifton** [comment]
No Shopify use HMAC use throughout their API and whenever interacting with your app.

**Comment by froatsnook** [comment]
@lukeclifton I just did a countOrders request, and there was no X-Shopify-Hmac-SHA256 header:

```
{ server: 'nginx',
date: 'Thu, 17 Dec 2015 18:47:51 GMT',
'content-type': 'application/json; charset=utf-8',
'transfer-encoding': 'chunked',
connection: 'keep-alive',
vary: 'Accept-Encoding',
status: '200 OK',
'x-frame-options': 'DENY',
'x-shopid': 'XXXXXXX',
'x-shardid': 'X',
'x-shopify-shop-api-call-limit': '1/40',
http_x_shopify_shop_api_call_limit: '1/40',
'x-stats-userid': '0',
'x-stats-apiclientid': 'XXXXXX',
'x-stats-apipermissionid': 'XXXXXXX',
'x-xss-protection': '1; mode=block; report=/xss-report/XXXXXXXXXXXXXXX?source%5Baction%5D=
count&source%5Bcontroller%5D=admin%2Forders&source%5Bsection%5D=admin',
'x-request-id': 'XXXXXXXXXXXXXXXXXXXXXXXX',
p3p: 'CP="NOI DSP COR NID ADMa OPTa OUR NOR"',
'x-dc': 'ash',
'x-content-type-options': 'nosniff' },
```

I mean, it only makes sense to me anyway for webhooks, since no one else should have a valid https certificate for *.myshopify.com.  Are you seeing it with all replies from shopify?  Or just incoming requests?

**Comment by lukeclifton** [comment]
Its in the form of a url query param. Its included along with the shop and signature params. Its useful particularly during the oAuth process so you can ensure its coming from Shopify.

Also useful for when running an embedded app.

**Comment by froatsnook** [comment]
@lukeclifton Got it.  Thanks for the info.

**Comment by froatsnook** [comment]
For now I'm not planning on implementing keyset persistence.  I think this should usually live in users.  A ShopifyCredentials collection could be implemented in a package, though!
